### ISSUE TITLE
iOS 15.5 libMobileGestalt.dylib disassembly issues

### ISSUE THREAD

**Issue by PoomSmart** [arandomdev/DyldExtractor#40]
**Describe the bug**

Dead addresses for strings section:

![image](https://user-images.githubusercontent.com/3608783/169645039-df908ff2-af54-4f81-877b-32b0cad0a589.png)

Lots of strings (obfuscated MobileGestalt keys) are not referenced (compared to the x86_64 semi-equivalent of the binary):

![image](https://user-images.githubusercontent.com/3608783/169645097-380b5b3e-59d5-4ed9-a668-d7d1f2e999b5.png)

Some instructions failed to be disassembled correctly (`1910891C0` for example):

![image](https://user-images.githubusercontent.com/3608783/169645184-eb22f398-fd10-4b1f-8f70-e08b34e80ceb.png)

**To Reproduce**
1. iOS 15.5
2. dyld_shared_cache_arm64 of **iPod9,1** (https://www.theiphonewiki.com/wiki/Firmware/iPod_touch/15.x)
3. Terminal command to reproduce.

**Additional context**

**Comment by arandomdev** [comment]
Okay, as for the dead address, 0x1F4136FB8, that is an `isa` that points to ___CFConstantStringClassReference, this shouldn't effect anything, but I can look into pulling the class in much like how I would pull in a super classes.

As for the other two issues, I'm kind of at a lost, and can only blame `libMobileGestalt.dylib`.  Before any processing, I found that the CFStrings are already like that. I do see some obfuscated strings, so maybe its just logging or unofuscated information that's mixed in?

And for that spot in the `__text`, it's originally just all zeros...?

**Comment by PoomSmart** [comment]
@arandomdev Alright, for some strings not being referenced, take `XnXl4MhKZx3zRKvA7ZwIYQ` as an example.

On x86_64, it is referenced in conjunction with `sub_AD8C`.

![image](https://user-images.githubusercontent.com/3608783/169678455-154c22ec-d536-43bd-87d6-3a621f5f0109.png)

Inside `sub_AD8C`, it calls an internal function that I named as `GetDeviceClass`:

![image](https://user-images.githubusercontent.com/3608783/169678479-63873985-14d0-42c7-9411-7341cb445ede.png)
![image](https://user-images.githubusercontent.com/3608783/169678505-a4f84d54-2177-41f0-a485-10ac069daf6c.png)

Which leads us to another internal function that I named `GetDeviceClassInternal`.
![image](https://user-images.githubusercontent.com/3608783/169678521-35ed03bb-808a-44c5-9a8c-31b19b157a58.png)

This same `GetDeviceClassInternal` exists in ARM version of `libMobileGestalt.dylib` but there's no such reference to it.

![image](https://user-images.githubusercontent.com/3608783/169678550-70b52cc8-e62f-47ef-9dd2-50008bed11fb.png)

**Comment by PoomSmart** [comment]
For the spot in `__text` that you assumed to be zero, I will try to figure out what they are in x86_64 version of them.

**Comment by PoomSmart** [comment]
Here, I compare `GetDeviceClass` function between iOS 15.5 ARM and iOS 14.6 ARM

iOS 15.5 ARM (take a look at `unk_1EBEBDA78`):

![image](https://user-images.githubusercontent.com/3608783/169678793-e25f5603-65e3-45ba-836f-45b1b4cfdb89.png)

![image](https://user-images.githubusercontent.com/3608783/169678808-47da8b17-6baa-4e41-8a6a-1f234b57be14.png)

iOS 14.6 ARM (take a look at `stru_1DB800420`)

![image](https://user-images.githubusercontent.com/3608783/169678826-83fc6a98-02c4-4ed7-811d-60ada6f253f7.png)

![image](https://user-images.githubusercontent.com/3608783/169678843-540bb522-4282-4e89-9758-6d11d42c01a6.png)

iOS 14.6 one gets more complete information.

**Comment by arandomdev** [comment]
Regarding `XnXl4MhKZx3zRKvA7ZwIYQ`, I was able to find it in the ARM version as well.
![image](https://user-images.githubusercontent.com/40927866/169708126-3fa7dd5c-7dca-4b5c-8580-26f1c3ebdef0.png)
![image](https://user-images.githubusercontent.com/40927866/169708141-845482b7-2571-46d4-9fac-f3e427fe7945.png)
![image](https://user-images.githubusercontent.com/40927866/169708156-432eb856-7d21-4e87-acab-ce04e866cbe7.png)
![image](https://user-images.githubusercontent.com/40927866/169708174-c84e3270-20e7-4835-8315-df1ae5001ba8.png)
I did this by first searching for the string, and then searching for the pointer to the string, which bought me to `0x1ebecf288`.

For the block layout, it seems to be the same issue with the `isa`. In 14.6, it probably has the dyld_info_command with binding info, which allows your decompiler to override the old address (`0x1f4130b58`), with a generated stub address for `__NSConcreteGlobalBlock` and continue from there. Ideally I we could generating new binding info for newer images that don't have it, but that's been scoped for DyldExtractorC. Alternatively, we can try to pull in the `isa`, but that might be challenging because the block layouts are spread out in the `__const` section, and we wouldn't know if they're there or where are they.
